mbrl/util/common.py

```diff
@@ -190,6 +194,116 @@ def create_replay_buffer(
+        and validation iterators, respectively.
+    """
+    data = replay_buffer.get_all(shuffle=True)
+    val_size = int(replay_buffer.num_stored * val_ratio)
+    train_size = replay_buffer.num_stored - val_size
+    train_data = data[:train_size]
+    train_iter = BootstrapIterator(
+        train_data,
+        batch_size,
+        ensemble_size,
+        shuffle_each_epoch=shuffle_each_epoch,
+        permute_indices=bootstrap_permutes,
+        rng=replay_buffer._rng,
+    )
+
+    val_iter = None
+    if val_size > 0:
+        val_data = data[train_size:]
+        val_iter = TransitionIterator(
+            val_data, batch_size, shuffle_each_epoch=False, rng=replay_buffer._rng
+        )
+
+    return train_iter, val_iter
+
+
+def get_sequence_buffer_iterator(
+    replay_buffer: ReplayBuffer,
+    batch_size: int,
+    val_ratio: float,
+    sequence_length: int,
+    ensemble_size: Optional[int] = None,
+    max_batches_per_loop: Optional[int] = None,
+) -> Tuple[SequenceTransitionIterator, Optional[SequenceTransitionIterator]]:
+    """Returns training/validation iterators for the data in the replay buffer.
+
+    Args:
+        replay_buffer (:class:`mbrl.util.ReplayBuffer`): the replay buffer from which
+            data will be sampled.
+        batch_size (int): the batch size for the iterators.
+        val_ratio (float): the proportion of data to use for validation. If 0., the
+            validation buffer will be set to ``None``.
+        sequence_length (int): the length of the sequences returned by the iterators.
+        ensemble_size (int): the number of models in the ensemble.
+        max_batches_per_loop (int, optional): if given, specifies how many batches
+            to return (at most) over a full loop of the iterator.
+
+    Returns:
+        (tuple of :class:`mbrl.replay_buffer.SequenceTransitionIterator`): the training
+        and validation iterators, respectively.
+    """
+
+    transitions = replay_buffer.get_all()
+    num_trajectories = len(replay_buffer.trajectory_indices)
+    val_size = int(num_trajectories * val_ratio)
+    train_size = num_trajectories - val_size
+    train_trajectories = replay_buffer.trajectory_indices[:train_size]
+
+    train_iterator = SequenceTransitionIterator(
+        transitions,
+        train_trajectories,
+        batch_size,
+        sequence_length,
+        ensemble_size,
+        rng=replay_buffer._rng,
+        max_batches_per_loop=max_batches_per_loop,
+    )
+
+    val_iterator = None
+    if val_size > 0:
+        val_trajectories = replay_buffer.trajectory_indices[train_size:]
+        val_iterator = SequenceTransitionIterator(
+            transitions,
+            val_trajectories,
+            batch_size,
+            sequence_length,
+            1,
+            rng=replay_buffer._rng,
+            max_batches_per_loop=max_batches_per_loop,
+        )
+        val_iterator.toggle_bootstrap()
```

Comment:
True, it's not really necessary since in the latest version the iterator always sets `it._bootstrap_iter = False` if the ensemble size is 1. But, unless I'm missing something , this toggle shouldn't change anything, I'm curious why you had to comment it out.

Reply:
I tried to reuse the evaluate function for training and needed to disable both. The double toggle made it weird.
I think this should be already in the latest commit in #88 
```python
   def evaluate(self, dataset: TransitionIterator) -> torch.Tensor:
        """Evaluates the model on the validation dataset.

        Iterates over the dataset, one batch at a time, and calls
        :meth:`mbrl.models.Model.eval_score` to compute the model score
        over the batch. The method returns the average score over the whole dataset.

        Args:
            dataset (bool): the transition iterator to use.

        Returns:
            (tensor): The average score of the model over the dataset (and for ensembles, per
                ensemble member).
        """
        if isinstance(dataset, BootstrapIterator):
            dataset.toggle_bootstrap()
        if isinstance(dataset, SequenceTransitionIterator):
            dataset.toggle_sequenced()

        batch_scores_list = []
        for batch in dataset:
            avg_batch_score = self.model.eval_score(batch)
            batch_scores_list.append(avg_batch_score)
        batch_scores = torch.cat(batch_scores_list, axis=batch_scores_list[0].ndim - 2)

        if isinstance(dataset, BootstrapIterator):
            dataset.toggle_bootstrap()
        if isinstance(dataset, SequenceTransitionIterator):
            dataset.toggle_sequenced()

        mean_axis = 1 if batch_scores.ndim == 2 else (1, 2)
        batch_scores = batch_scores.mean(axis=mean_axis)

        return batch_scores`

Reply:
Do you get an error if you don't remove the toggle from the common utility? If so, can you paste the error message here? I'm trying to figure out if there is a bug, since I would have expected toggle to do nothing for the validation iterator with ensemble size 1.

Reply:
If you go to #88, use
`algorithm=pddm dynamics_model=gaussian_mmlp_ensemble overrides=pddm_cartpole overrides.validation_ratio=0.2` settings and comment in the toggle you should see something similar to
`...
 File "/home/fiber/projects/mbrl-lib/mbrl/models/gaussian_mlp.py", line 340, in eval_score
    assert model_in.ndim == 2 and target.ndim == 2
AssertionError`

Reply:
Oh, it looks like your version of `BootstrapIterator` is not up to date, check out [this](https://github.com/facebookresearch/mbrl-lib/blob/sequence_iterator/mbrl/util/replay_buffer.py#L170) line, which is missing in #88. While we finalize this PR, can you make the target of #88 be the `sequence_iterator` branch to make it easier to spot any discrepancies?

Reply:
Ok my fault, got it running locally